src/block.rs

```diff
@@ -1,140 +1,191 @@
+    Public,
+}
+
+// TODO: move to tiny_cid
+fn create_cid<M: MultihashDigest>(ccode: u64, hcode: u64, payload: &[u8]) -> Result<Cid> {
+    let digest = M::new(hcode, payload)
+        .map_err(|_| UnsupportedMultihash(hcode))?
+        .to_raw()
+        .map_err(|_| UnsupportedMultihash(hcode))?;
+    Ok(Cid::new_v1(ccode, digest))
 }
 
-/// Raw decode.
-///
-/// Useful for nested encodings when for example the data is encrypted.
-pub fn raw_decode<C, H, O, D>(codec: C, mut data: &[u8]) -> Result<D>
-where
-    O: Codec<C>,
-    D: Decode<O, C>,
-    C: Into<u64> + TryFrom<u64> + Copy + PartialEq,
-    H: Into<u64> + TryFrom<u64> + Copy,
-{
-    if codec != O::CODE {
-        return Err(Error::UnsupportedCodec(codec.into()));
+// TODO: move to tiny_cid
+fn verify_cid<M: MultihashDigest>(cid: &Cid, payload: &[u8]) -> Result<()> {
+    let hcode = cid.hash().code();
+    let mh = M::new(hcode, payload).map_err(|_| UnsupportedMultihash(hcode))?;
+    if mh.digest() != cid.hash().digest() {
+        return Err(InvalidMultihash(mh.to_bytes()).into());
     }
-    D::decode(&mut data).map_err(|e| Error::CodecError(Box::new(e)))
+    Ok(())
 }
 
-/// Decodes a block.
-pub fn decode<C, H, O, D>(cid: &CidGeneric<C, H>, data: &[u8]) -> Result<D>
-where
-    O: Codec<C>,
-    D: Decode<O, C>,
-    C: Into<u64> + TryFrom<u64> + Copy + PartialEq,
-    H: Into<u64> + TryFrom<u64> + Copy + PartialEq,
-    Box<dyn MultihashDigest<H>>: From<H>,
-{
-    if data.len() > MAX_BLOCK_SIZE {
-        return Err(Error::BlockTooLarge(data.len()));
+impl<C: Codec, M: MultihashDigest> Block<C, M> {
+    /// Creates a new block.
+    pub fn new(cid: Cid, data: Box<[u8]>) -> Self {
+        Self {
+            _marker: PhantomData,
+            cid,
+            data,
+            vis: Visibility::Public,
+        }
     }
-    let hash = Box::<dyn MultihashDigest<H>>::from(cid.hash().algorithm()).digest(&data);
-    if hash.as_ref() != cid.hash() {
-        return Err(Error::InvalidHash(hash.to_vec()));
+
+    /// Returns the blocks visibility.
+    pub fn visibility(&self) -> Visibility {
+        self.vis
     }
-    raw_decode::<C, H, O, D>(cid.codec(), data)
-}
 
-/// Decode block to ipld.
-pub fn decode_ipld<C, H>(cid: &CidGeneric<C, H>, data: &[u8]) -> Result<Ipld<C, H>>
-where
-    C: Into<u64> + TryFrom<u64> + Copy + PartialEq + EncodeDecodeIpld<H>,
-    H: Into<u64> + TryFrom<u64> + Copy + PartialEq,
-    Box<dyn MultihashDigest<H>>: From<H>,
-{
-    if data.len() > MAX_BLOCK_SIZE {
-        return Err(Error::BlockTooLarge(data.len()));
+    /// Sets the blocks visibility.
+    pub fn set_visibility(&mut self, vis: Visibility) {
+        self.vis = vis;
     }
-    let hash = Box::<dyn MultihashDigest<H>>::from(cid.hash().algorithm()).digest(&data);
-    if hash.as_ref() != cid.hash() {
-        return Err(Error::InvalidHash(hash.to_vec()));
+
+    /// Encode a block.`
+    pub fn encode<T: Encode<C> + ?Sized>(codec: C, hcode: u64, payload: &T) -> Result<Self> {
```

Comment:
Is there a way to unify `encode()` and `encode_ipld()`. I haven't looked deep enough, I just wanted to post this question to make sure I don't forget to dig deeper into that. I'm even not sure if unifying them would be a good idea even if it's possible.

Reply:
This is not unifiable as it would require expressing that Ipld implements Encode<C>. Rust doesn't support trait bounds on concrete types.